### ISSUE TITLE
IndexFailedEngineException[[encrypt_idx2][2] Index failed for [test_doc#32]]; nested: HttpHostConnectException[Connect to 0.0.0.0:80 [/0.0.0.0] failed: Operation timed out (Connection timed out)];

### ISSUE THREAD

**Issue by checkking** [Workday/elasticrypt#9]
When I PUT create and doc, an error happened. Can you tell me why.

**Comment by ilvar** [comment]
Looks like you haven't set up a key provider. By default, we're using `HttpKeyProvider` that gets keys over HTTP interface, so you need to build that part yourself. Or write your own provider.

**Comment by checkking** [comment]
@ilvar  By the way, Is this project has been used in production, and is there any side-effects?

**Comment by ilvar** [comment]
Yes, it is used in production. Obviously, disk IO is significantly slower, as we need to encrypt/decrypt every page, even if it is cached in memory by the OS.

**Comment by checkking** [comment]
@ilvar Do you mean you encrypt the data by page? Why do not encrypt the whole data, and then write it to disk by page?

**Comment by checkking** [comment]
There is a setting, **"index.store.type": "encrypted"**, but I do not see **encrypted** in source code, how is  **"index.store.type": "encrypted"** effective.

**Comment by ilvar** [comment]
ES never reads or writes the whole data, dataset for just one shard can easily be larger than the whole Java heap available (32 gb max).

On version: our code is closely coupled with low level Lucene disk io, so it won't even compile if you just bump es (and Lucene) versions. Though, I don't think they've changed the whole interface, so probably the code change necessary wouldn't be very sophisticated. Feel free to fork and try!

**Comment by ilvar** [comment]
I'm not 100% sure on settings but looks like it picks module at the `org/elasticsearch/index/store/EncryptedDirectoryService.scala` with `Encrypted`, the same with other store types.
